calico_node/startup.go

```diff
@@ -14,59 +14,657 @@
+	}
+
+	// Set other Felix config that is not yet in the node resource.
+	if err := ensureDefaultConfig(client, node); err != nil {
+		fatal("Unable to set global default configuration: %s", err)
+		terminate()
+	}
+
+	// Write the startup.env file now that we are ready to start other
+	// components.
+	ipv6Str := ""
+	if node.Spec.BGP.IPv6Address != nil {
+		ipv6Str = node.Spec.BGP.IPv6Address.String()
+	}
+	writeStartupEnv(nodeName, node.Spec.BGP.IPv4Address.String(), ipv6Str)
+
+	// Tell the user what the name of the node is.
+	message("Using node name: %s", nodeName)
+}
+
+// determineNodeName is called to determine the node name to use for this instance
+// of calico/node.
+func determineNodeName() string {
+	// Determine the name of this node.
+	nodeName := os.Getenv("NODENAME")
+	if nodeName == "" {
+		// NODENAME not specified, check HOSTNAME (we maintain this for
+		// backwards compatibility).
+		log.Info("NODENAME environment not specified - check HOSTNAME")
+		nodeName = os.Getenv("HOSTNAME")
+	}
+	if nodeName == "" {
+		// The node name has not been specified.  We need to use the OS
+		// hostname - but should warn the user that this is not a
+		// recommended way to start the node container.
+		var err error
+		if nodeName, err = os.Hostname(); err != nil {
+			log.Info("Unable to determine hostname - exiting")
+			panic(err)
+		}
+
+		message("******************************************************************************")
+		message("* WARNING                                                                    *")
+		message("* Auto-detecting node name.  It is recommended that an explicit fixed value  *")
+		message("* is supplied using the NODENAME environment variable.  Using a fixed value  *")
+		message("* ensures that any changes to the compute host's hostname will not affect    *")
+		message("* the Calico configuration when the Calico node restarts.                    *")
+		message("******************************************************************************")
+	}
+	return nodeName
+}
+
+// createClient() creates the Calico API client, and unless requested otherwise,
+// checks that the client is accessible.
+func createClient() (*api.CalicoAPIConfig, *client.Client) {
+	log.Info("Loading client config")
 	cfg, err := client.LoadClientConfig("")
 	if err != nil {
-		panic(fmt.Sprintf("Error loading config: %s", err))
+		fatal("Error loading datastore config: %s", err)
+		terminate()
 	}
-	c, err := backend.NewClient(*cfg)
+
+	log.Info("Creating client")
+	client, err := client.New(*cfg)
 	if err != nil {
-		panic(fmt.Sprintf("Error creating client: %s", err))
+		fatal("Error accessing the Calico datastore: %s", err)
+		terminate()
+	}
+
+	// An explicit value of false is required to skip waiting for the
+	// datastore.
+	if os.Getenv("WAIT_FOR_DATASTORE") == "false" {
+		message("Skipping datastore connection test")
+		return cfg, client
+	}
+
+	message("Checking datastore connection")
+	for i := 0; i <= 120; i++ {
```

Comment:
I just couldn't bring myself to make it loop indefinitel, but you're probably right let's keep the scope tight.